src/fast_array_utils/stats/_sum.py

```diff
@@ -69,39 +96,73 @@ def _sum_dask(
         msg = "sum does not support numpy matrices"
         raise TypeError(msg)
 
-    def sum_drop_keepdims(
-        a: CpuArray,
+    def sum_dask_inner(
+        a: CpuArray | GpuArray,
         /,
         *,
-        axis: tuple[Literal[0], Literal[1]] | Literal[0, 1, None] = None,
+        axis: ComplexAxis = None,
         dtype: DTypeLike | None = None,
         keepdims: bool = False,
     ) -> NDArray[Any] | types.CupyArray:
-        del keepdims
-        if a.ndim == 1:
-            axis = None
-        else:
-            match axis:
-                case (0, 1) | (1, 0):
-                    axis = None
-                case (0 | 1 as n,):
-                    axis = n
-                case tuple():  # pragma: no cover
-                    msg = f"`sum` can only sum over `axis=0|1|(0,1)` but got {axis} instead"
-                    raise ValueError(msg)
-        rv = sum(a, axis=axis, dtype=dtype)
-        shape = (1,) if a.ndim == 1 else (1, 1 if rv.shape == () else len(rv))  # type: ignore[arg-type]
-        return np.reshape(rv, shape)
+        axis = normalize_axis(axis, a.ndim)
+        rv = sum(a, axis=axis, dtype=dtype, keep_cupy_as_array=True)  # type: ignore[misc,arg-type]
+        shape = get_shape(rv, axis=axis, keepdims=keepdims)
+        return cast("NDArray[Any] | types.CupyArray", rv.reshape(shape))
 
     if dtype is None:
         # Explicitly use numpy result dtype (e.g. `NDArray[bool].sum().dtype == int64`)
         dtype = np.zeros(1, dtype=x.dtype).sum().dtype
 
-    return da.reduction(
+    rv = da.reduction(
         x,
-        sum_drop_keepdims,  # type: ignore[arg-type]
-        partial(np.sum, dtype=dtype),  # pyright: ignore[reportArgumentType]
+        sum_dask_inner,  # type: ignore[arg-type]
+        partial(sum_dask_inner, dtype=dtype),  # pyright: ignore[reportArgumentType]
         axis=axis,
         dtype=dtype,
         meta=np.array([], dtype=dtype),
     )
+    if (
+        axis is None
+        and not keep_cupy_as_array
+        and isinstance(x._meta, types.CupyArray | types.CupyCSMatrix)  # noqa: SLF001
+    ):
+
+        def to_scalar(a: types.CupyArray) -> np.number[Any]:
+            return cast("np.number[Any]", a.get()[()])
+
+        return rv.map_blocks(to_scalar, meta=x.dtype.type(0))  # type: ignore[arg-type]
+    return rv
+
+
+def normalize_axis(axis: ComplexAxis, ndim: int) -> Literal[0, 1, None]:
+    """Adapt `axis` parameter passed by Dask to what we support."""
+    match axis:
+        case int() | None:
+            pass
+        case (0 | 1,):
+            axis = axis[0]
+        case (0, 1) | (1, 0):
+            axis = None
+        case _:  # pragma: no cover
+            raise AxisError(axis, ndim)  # type: ignore[call-overload]
+    if axis == 0 and ndim == 1:
+        return None  # dask’s aggregate doesn’t know we don’t accept `axis=0` for 1D arrays
```

Comment:
This is the fix? Everything else is refactor?

Reply:
Everything’s messy because cupy and numpy have subtly different behavior. So if the shapes aren’t exactly right in all cases, something blows up (Case in point: this PR currently has a regression for `axis=None` that I need to fix)

Reply:
Got it I see your above comment.